[release/3.1] Fix corefx to build on clang 10

omajid opened the pull request:
Clang 10 adds/enables new warnings, some of which is affecting the corefx code.

Clang 10 has added `-Walloca` to warn about uses of `alloca()`. Clang 10 has also added `-Wimplicit-int-float-conversion`. This commit disables those warnings where applicable.

This is a backport of dotnet/runtime#33734 to corefx.

After this commit, I can build all of corefx with Clang 10.

omajid commented:
cc @jkotas @lpereira @stephentoub

lpereira commented on src/Native/Unix/System.Native/pal_io.c:
```diff
@@ -906,18 +906,20 @@ int32_t SystemNative_Poll(PollEvent* pollEvents, uint32_t eventCount, int32_t mi
+    if (useStackBuffer)
     {
-        return SystemNative_ConvertErrorPlatformToPal(EOVERFLOW);
+        pollfds = (struct pollfd*)&stackBuffer[0];
```
```suggestion
        pollfds = &stackBuffer[0];
```

lpereira commented on src/Native/Unix/System.Native/pal_io.c:
```diff
@@ -906,18 +906,20 @@ int32_t SystemNative_Poll(PollEvent* pollEvents, uint32_t eventCount, int32_t mi
+    else
     {
-        return Error_ENOMEM;
+        pollfds = (struct pollfd*)calloc(eventCount, sizeof(*pollfds));
```
```suggestion
        pollfds = calloc(eventCount, sizeof(*pollfds));
```

lpereira commented on src/Native/Unix/System.Native/pal_io.c:
```diff
@@ -906,18 +906,20 @@ int32_t SystemNative_Poll(PollEvent* pollEvents, uint32_t eventCount, int32_t mi
-    size_t bufferSize;
-    if (!multiply_s(sizeof(struct pollfd), (size_t)eventCount, &bufferSize))
+    struct pollfd stackBuffer[(uint32_t)(2048/sizeof(struct pollfd))];
+    int useStackBuffer = eventCount <= (sizeof(stackBuffer)/sizeof(stackBuffer[0]));
```
```suggestion
    int useStackBuffer = eventCount <= ARRAY_SIZE(stackBuffer);
```

lpereira reviewed (approved):
LGTM with suggestions

omajid commented:
@lpereira 

> LGTM with suggestions

Should I make these suggestions in `runtime` too? This same PR was applied (and merged) into runtime already: https://github.com/dotnet/runtime/pull/33734

jkotas commented:
Is this just fixing warnings? For servicing, we should disable the warnings instead of touching the code.

jkotas commented:
cc @danmosemsft

danmoseley commented:
If we can dial this back to disabling warnings, I am happy to seek approval to merge it.

omajid committed:
Fix corefx to build on clang 10

Clang 10 adds/enables new warnings, some of which is affecting the
corefx code. Disable them.

This is a "backport" of dotnet/runtime#33734 to corefx.

After this commit, I can build all of corefx with Clang 10.

omajid commented on src/Native/Unix/CMakeLists.txt:
```diff
@@ -23,6 +23,10 @@ add_compile_options(-I${CMAKE_CURRENT_SOURCE_DIR}/Common)
 add_compile_options(-I${CMAKE_CURRENT_BINARY_DIR}/Common)
 add_compile_options(-g)
 if(CMAKE_C_COMPILER_ID STREQUAL Clang)
+    if (CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 10)
```
I am not a fan of checking the compiler version. It would be better to check if the compiler supports this flag at all. But I dont see us doing this for any other compiler flag.

jkotas commented:
@danmosemsft This is dialed back to disabling warnings now. Could you please get it approved?

danmoseley commented:
Yup - @omajid can you share why you're working to enable build with clang 10 ? Does RedHat plans to build its 3.1 servicing releases using clang 10?

omajid commented:
>  Does RedHat plans to build its 3.1 servicing releases using clang 10?

I am not aware of any concrete plans for RHEL.

> can you share why you're working to enable build with clang 10 ?

My main motivation is to build .NET Core in Fedora (>= 32) where clang is at version 10. We are currently carrying a local patch for that: https://src.fedoraproject.org/fork/crummel/rpms/dotnet3.1/blob/master/f/corefx-42900-clang-10.patch. I prefer to avoid carrying these patches out of band where possible.

danmoseley commented:
@Anipik will merge when branch open
